components/rendering/src/shortcode.rs

```diff
@@ -137,8 +140,22 @@ fn render_shortcode(
     }
 }
 
+struct InvocationTracker(HashMap<String, u32>);
+
+impl InvocationTracker {
+    fn new() -> Self {
+        InvocationTracker(HashMap::new())
+    }
+    fn get(&mut self, name: &str) -> u32 {
+        let invocation_number = self.0.entry(String::from(name)).or_insert(0);
+        *invocation_number += 1;
+        *invocation_number
+    }
+}
+
 pub fn render_shortcodes(content: &str, context: &RenderContext) -> Result<String> {
     let mut res = String::with_capacity(content.len());
+    let mut invocation_counts = InvocationTracker::new();
```

Comment:
why a struct and not simply a hashmap?

Reply:
The intent was to simplify code at the call site by wrapping the lazy-initialization and increment logic in the `get()` method. Would it be better to do something like:
```rust
    let mut invocation_map: HashMap<String, u32> = HashMap::new();
    let mut get_invocation_count = |name: &str| {
        let invocation_number = invocation_map.entry(String::from(name)).or_insert(0);
        *invocation_number += 1;
        *invocation_number
    };
```
instead?

Reply:
Yep